ingest-router/src/project_config/handler.rs

```diff
@@ -0,0 +1,396 @@
+
+#[async_trait]
+impl Handler<ProjectConfigsRequest, ProjectConfigsResponse> for ProjectConfigsHandler {
+    /// Pending public keys that couldn't be routed to any cell
+    type SplitMetadata = Vec<String>;
+
+    async fn split_requests(
+        &self,
+        request: ProjectConfigsRequest,
+        _cells: &Cells,
+    ) -> Result<(Vec<(CellId, ProjectConfigsRequest)>, Vec<String>), IngestRouterError> {
+        let public_keys = request.public_keys;
+        let extra_fields = request.extra_fields;
+
+        // Route each public key to its owning cell using the locator service
+        let mut split: HashMap<CellId, Vec<String>> = HashMap::new();
+        let mut pending: Vec<String> = Vec::new();
+
+        for public_key in public_keys {
+            match self.locator.lookup(&public_key, None).await {
+                Ok(cell_id) => {
+                    split.entry(cell_id).or_default().push(public_key);
+                }
+                Err(e) => {
+                    // Locator errors, add to pending
+                    tracing::error!(
+                        public_key = %public_key,
+                        error = ?e,
+                        "Failed to route public key"
+                    );
+                    pending.push(public_key);
+                }
+            }
+        }
+
+        // Build per-cell requests
+        let cell_requests: Vec<(CellId, ProjectConfigsRequest)> = split
+            .into_iter()
+            .map(|(cell_id, keys)| {
+                (
+                    cell_id,
+                    ProjectConfigsRequest {
+                        public_keys: keys,
+                        extra_fields: extra_fields.clone(),
+                    },
+                )
+            })
+            .collect();
+
+        Ok((cell_requests, pending))
+    }
+
+    fn merge_results(
+        &self,
+        results: Vec<Result<(CellId, ProjectConfigsResponse), (CellId, IngestRouterError)>>,
+        pending_from_split: Vec<String>,
+    ) -> ProjectConfigsResponse {
+        let mut merged = ProjectConfigsResponse::new();
+
+        // Add pending keys from split phase
+        if !pending_from_split.is_empty() {
+            merged.pending_keys.extend(pending_from_split);
+        }
+
+        // Results are provided pre-sorted by cell priority (highest first)
+        // The executor ensures results are ordered so we can use the first successful response
+        // for extra_fields and headers.
+        // Failed cells are handled by the executor adding their keys to pending_from_split.
+        let mut found_priority_cell = false;
+
+        // Process successful results from each cell (already in priority order)
+        for result in results.into_iter().flatten() {
+            let (_cell_id, response) = result;
+
+            if !response.project_configs.is_empty() {
+                merged
+                    .project_configs
+                    .extend(response.project_configs.clone());
```

Comment:
are these clones needed? it seems like the response data isn't needed elsewhere so we should be able to avoid copying bytes.